### PR TITLE
My Site Dashboard: use diffable data source on the posts card

### PR DESCRIPTION
**Pull request by leandroalonso**
Fixes one last remaining issue in #17697 and improves #17875

This PR makes two changes:

* Make use of diffable data source instead of adding/deleting/updating cells the old way — allowing for a clean code and the removal of a hack to keep 3 posts being displayed.
* Fixes a bug in which the post card had a minimum height

https://user-images.githubusercontent.com/7040243/159077288-88729f8b-1572-4264-a95b-d415613c582e.mp4

### To test

Enable My Site Dashboard flag and:

#### Discarding a new post

1. Check the dashboard for a site with 3 or more draft posts
2. Tap the FAB
3. Create a new post
4. Tap X
5. ✅ The number of posts displayed should be 3

#### Displaying the correct height

https://user-images.githubusercontent.com/7040243/159077787-7cc5935e-6714-4183-a567-709110a3caa0.mp4

1. Make sure to have a site with no posts
2. Close the app
3. Open it
4. When the dashboard appears the cell should not jump
5. ✅ Create a new post (using the FAB)
6. Save as draft
7. ✅ Make sure the height of the card is correct

#### Ensuring a max of 3 posts

https://user-images.githubusercontent.com/7040243/159077927-3998fd1e-183b-4515-abdd-cee857b1c873.mp4

1. Make sure to have a site with more than 4 draft posts
2. Open the dashboard
5. Go to Drafts
6. Go back to the dashboard
7. ✅ The number of posts displayed shouldn't change

## Regression Notes
1. Potential unintended areas of impact
The currently posts card.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Tested as described above.

3. What automated tests I added (or what prevented me from doing so)
n/a

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Posts/PostsCardViewModel.swift [hunk 1]
```diff
@@ -35,6 +35,21 @@ class PostsCardViewModel: NSObject {
 
     private weak var viewController: PostsCardView?
 
+    lazy var diffableDataSource = UITableViewDiffableDataSource<Int, NSManagedObjectID>(tableView: viewController!.tableView) { [weak self] (tableView, indexPath, objectID) -> UITableViewCell? in
```
Comment (hassaanelgarem) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Posts/PostsCardViewModel.swift]: [nit] You could add a couple of type aliases to clean up the code a bit:
```
typealias DataSource = UITableViewDiffableDataSource<Int, NSManagedObjectID>
typealias Snapshot = NSDiffableDataSourceSnapshot<Int, NSManagedObjectID>
```

**Review by hassaanelgarem — COMMENTED** [review]
Everything works as described except for the card height issue. I'm still experiencing a slight jump, however, I feel it's better than before:

https://user-images.githubusercontent.com/25306722/159199646-c7647f21-8e66-45ff-8cde-332c433d7ef8.mp4

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Posts/PostsCardViewModel.swift (hunk 1 above)
Comment (leandroalonso) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Posts/PostsCardViewModel.swift]: Done in 3b3dfd6b4e033ec8be565623fb761af71777b256

**Comment by leandroalonso** [comment]
> Everything works as described except for the card height issue. I'm still experiencing a slight jump, however, I feel it's better than before:

@hassaanelgarem this is expected. I didn't found a good fix for this yet. What happens is:

* We know from the API if we should show a next post prompt
* However, the user might have a local post
* Given the above, we perform a query on the database to see if there's any post
* When the response arrives we then show the prompt or the post(s)

The layout calculation for the collection view is faster than this flow. Plus, we have two different prompts (for the first post and for the next post) so we set a height, but it might not be the correct one.

We can probably take a look at that in the future to improve it.

**Review by hassaanelgarem — APPROVED** [review]

**Review by momo-ozawa — COMMENTED** [review]
Hey @leandroalonso 👋 

Testing instruction work as described, but I found one issue. If you discard a post on a site with "Create your first post" or "Create your next post", the post card is displayed on top of the Quick Actions cards.

Create your first post | Create your next post
-- | --
 |  



https://user-images.githubusercontent.com/6711616/159417859-b4f1f3a7-94a0-4422-bd8d-c622b043e0fa.mp4

**Comment by leandroalonso** [comment]
@momo-ozawa fixed in 084b225e0877048d6021a3e0b6c654c5bdf5f77d

**Review by momo-ozawa — COMMENTED** [review]
@leandroalonso The [issue](https://github.com/wordpress-mobile/WordPress-iOS/pull/18165#pullrequestreview-916708499) I noted above about discarding posts is fixed. However, I'm seeing a new issue -- when switching from a "Create your first post" site to a "Create your next post" site (and vice versa), the following assertion failure occurs and the app crashes

```
        guard let dataSource = viewController?.tableView.dataSource as? DataSource else {
            assertionFailure("The data source has not implemented snapshot support while it should")
            return
        }
 ```

**Comment by leandroalonso** [comment]
@momo-ozawa I wasn't able to reproduce that. 🤔 

I removed the `assertionFailure`, could you please let me know if the behavior is now the expected?

Edit: I was able to reproduce that while switching sites. By removing it, the behavior is still correct.

**Review by momo-ozawa — APPROVED** [review]
> I removed the assertionFailure, could you please let me know if the behavior is now the expected?

Yes! Now it's behaving as expected

**Pull request merged**